### ISSUE TITLE
Cache only stored after 2nd query execution

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#13262]
**Describe the bug**
When caching is enabled, it takes two query executions before it's stored to cache.

**To Reproduce**
1. Native question > MySQL or Postgres or another database which supports a `sleep` function
```
SELECT pg_sleep(5) --Postgres
SELECT sleep(5) #MySQL
```
2. Save question as "Q1" and refresh page, you should see something like this in the log (5 seconds):
```
09-17 11:12:56 DEBUG middleware.log :: POST /api/card/183/query 202 [ASYNC: completed] 5.0 s (8 DB calls) App DB connections: 2/15 Jetty threads: 3/50 (1 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
```
3. Admin > Settings > Caching > enable caching Duration=1, TTL=10, Size=1000
4. Go to the saved question "Q1" and it will show the same log as in step 2 (5 seconds)
5. Refresh the page, which will now save the query to cache (5 seconds)
```
09-17 11:13:05 INFO api.card :: Question's average execution duration is 5.0 s; using 'magic' TTL of 5.8 days 💾
09-17 11:13:10 INFO middleware.cache :: Query took 5.0 s to run; miminum for cache eligibility is 1000.0 µs
09-17 11:13:10 INFO middleware.cache :: Caching results for next time for query with hash "5e28384f". 💾
09-17 11:13:10 DEBUG middleware.log :: POST /api/card/183/query 202 [ASYNC: completed] 5.0 s (10 DB calls) App DB connections: 1/15 Jetty threads: 3/50 (1 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
```
6. Refresh the page again, it will now load from cache (few milliseconds)
```
09-17 11:13:15 INFO api.card :: Question's average execution duration is 5.0 s; using 'magic' TTL of 5.8 days 💾
09-17 11:13:15 DEBUG middleware.log :: POST /api/card/183/query 202 [ASYNC: completed] 12.1 ms (5 DB calls) App DB connections: 1/15 Jetty threads: 3/50 (1 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
```

**Expected behavior**
Storing to cache on step 4 (first run) instead of step 5 (second run)

**Information about your Metabase Installation:**
Metabase 1.36.4, 0.36.6, 0.34.3, `master`

**Additional context**
Not sure if this somehow is aiding the problems in #13164

**Comment by camsaul** [comment]
Caching happens asynchronously this was probably fixed by #20310

**Comment by nemanjaglumac** [comment]
Github accidentally automatically closed this issue because I mentioned it in the PR, but I haven't used any of the closing keywords.

**Comment by flamber** [comment]
@nemanjaglumac https://github.com/metabase/metabase/commit/8e174f0bdc6e58f8b5aa5d80134230b3f5978e3a
`Note: should be revisited and enabled once we fix #13262.` - the trigger being `fix #13262`

**Comment by qnkhuat** [comment]
As of master(f16f2deff802c82933bd933183c40df759a6ea2b), cache hits the first time running question.
